File: packages/react-native/ios/Recording/ViewHierarchyScanner.swift
import UIKit

@objc public class ViewHierarchyScanner: NSObject {
    
    @objc public static let shared = ViewHierarchyScanner()
    @objc public var maxDepth: Int = 12
    @objc public var includeTextContent: Bool = true
    @objc public var includeVisualProperties: Bool = true
    
    private let _timeBudgetNs: UInt64 = 16_000_000
    
    private override init() {
        super.init()
    }
    
    @objc public func captureHierarchy() -> [String: Any]? {
        guard let w = _keyWindow() else { return nil }
        return serializeWindow(w)
    }
    
    @objc public func serializeWindow(_ window: UIWindow) -> [String: Any] {
        let ts = Int64(Date().timeIntervalSince1970 * 1000)
        let scale = window.screen.scale
        let bounds = window.bounds
        let start = DispatchTime.now().uptimeNanoseconds
        let root = _serializeView(window, depth: 0, start: start)
        
        var result: [String: Any] = [
            "timestamp": ts,
            "screen": ["width": bounds.width, "height": bounds.height, "scale": scale],
            "root": root ?? [:]
        ]
        
        if let sn = ReplayOrchestrator.shared.currentScreenName {
            result["screenName"] = sn
        }
        return result
    }
    
    private func _keyWindow() -> UIWindow? {
        if #available(iOS 15.0, *) {
            return UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }.flatMap { $0.windows }.first { $0.isKeyWindow }
        } else {
            return UIApplication.shared.windows.first { $0.isKeyWindow }
        }
    }
    
    private func _serializeView(_ view: UIView, depth: Int, start: UInt64) -> [String: Any]? {
        if depth > maxDepth { return nil }
        if (DispatchTime.now().uptimeNanoseconds - start) > _timeBudgetNs { return ["type": _typeName(view), "bailout": true] }
        if depth > 0 && (view.isHidden || view.alpha <= 0.01 || view.bounds.width <= 0 || view.bounds.height <= 0) { return nil }
        
        var node: [String: Any] = [:]
        node["type"] = _typeName(view)
        node["frame"] = ["x": view.frame.origin.x, "y": view.frame.origin.y, "w": view.frame.width, "h": view.frame.height]
        
        if view.isHidden { node["hidden"] = true }
        if view.alpha < 1.0 { node["alpha"] = view.alpha }
        if let aid = view.accessibilityIdentifier, !aid.isEmpty { node["testID"] = aid }
        if let lbl = view.accessibilityLabel, !lbl.isEmpty { node["label"] = lbl }
        if _isSensitive(view) { node["masked"] = true }
        
        if includeVisualProperties, let bg = view.backgroundColor, bg != .clear { node["bg"] = _hexColor(bg) }
        
        if includeTextContent {
            if let tv = view as? UITextView { node["text"] = _mask(tv.text ?? ""); node["textLength"] = tv.text?.count ?? 0 }
            else if let lb = view as? UILabel { node["text"] = _mask(lb.text ?? ""); node["textLength"] = lb.text?.count ?? 0 }
            else if let tf = view as? UITextField { node["text"] = "***"; node["textLength"] = tf.text?.count ?? 0; node["placeholder"] = tf.placeholder }
        }
        
        if _isInteractive(view) {
            node["interactive"] = true
            if let btn = view as? UIButton { if let t = btn.title(for: .normal) { node["buttonTitle"] = t }; node["enabled"] = btn.isEnabled }
            if let ctrl = view as? UIControl { node["enabled"] = ctrl.isEnabled }
        }
        
        if let sv = view as? UIScrollView {
            node["scrollEnabled"] = sv.isScrollEnabled
            node["contentOffset"] = ["x": sv.contentOffset.x, "y": sv.contentOffset.y]
            node["contentSize"] = ["w": sv.contentSize.width, "h": sv.contentSize.height]
        }
        
        if view is UIImageView { node["hasImage"] = true }
        
        let subs = view.subviews.filter { !$0.isHidden && $0.alpha > 0.01 }
        if !subs.isEmpty {
            var children: [[String: Any]] = []
            for s in subs {
                if let cn = _serializeView(s, depth: depth + 1, start: start) { children.append(cn) }
            }
            if !children.isEmpty { node["children"] = children }
        }
        
        return node
    }
    
    private func _typeName(_ v: UIView) -> String { String(describing: type(of: v)) }
    
    private func _isSensitive(_ v: UIView) -> Bool {
        if let tf = v as? UITextField, tf.isSecureTextEntry { return true }
        return false
    }
    
    private func _isInteractive(_ v: UIView) -> Bool {
        v is UIButton || v is UIControl || v is UITextField || v is UITextView
    }
    
    private func _mask(_ text: String) -> String {
        text.count > 100 ? String(text.prefix(100)) + "..." : text
    }
    
    private func _hexColor(_ c: UIColor) -> String {
        var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
        c.getRed(&r, green: &g, blue: &b, alpha: &a)
        return String(format: "#%02X%02X%02X", Int(r * 255), Int(g * 255), Int(b * 255))
    }
}

File: packages/rejourney/android/src/main/kotlin/com/rejourney/recording/ViewHierarchyScanner.kt
/**
 * Copyright 2026 Rejourney
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rejourney.recording

import android.app.Activity
import android.content.Context
import android.graphics.Rect
import android.os.SystemClock
import android.view.View
import android.view.ViewGroup
import android.widget.*
import java.lang.ref.WeakReference
import kotlin.math.roundToInt

/**
 * View hierarchy scanning and serialization
 * Android implementation aligned with iOS ViewHierarchyScanner.swift
 */
class ViewHierarchyScanner private constructor() {

    companion object {
        @Volatile
        private var instance: ViewHierarchyScanner? = null

        val shared: ViewHierarchyScanner
            get() = instance ?: synchronized(this) {
                instance ?: ViewHierarchyScanner().also { instance = it }
            }
    }

    /// Depth cap for the serialized tree.
    ///
    /// This was 12, and measurement showed it doing the cutting rather than
    /// protecting anything: across sampled production hierarchies 76% ended at
    /// exactly 12 -- the signature of truncation, not of trees that happen to
    /// be that deep -- while the 16ms time budget that is the real cost guard
    /// never fired once. Modern React Native trees nest well past 12 through
    /// wrapper views alone. The budget still bounds the work, and unlike this
    /// cap it marks what it cut.
    var maxDepth: Int = 24
    var includeTextContent: Boolean = true
    var includeVisualProperties: Boolean = true

    private val timeBudgetMs: Long = 16 // 16ms to stay under one frame

    private var currentActivity: WeakReference<Activity>? = null

    fun setCurrentActivity(activity: Activity?) {
        currentActivity = if (activity != null) WeakReference(activity) else null
    }

    fun captureHierarchy(): Map<String, Any>? {
        val activity = currentActivity?.get() ?: return null
        val decorView = activity.window?.decorView ?: return null
        return serializeWindow(decorView, activity)
    }

    fun serializeWindow(window: View, activity: Activity): Map<String, Any> {
        val ts = System.currentTimeMillis()
        val displayMetrics = activity.resources.displayMetrics
        val density = displayMetrics.density.takeIf { it > 0f } ?: 1f
        val bounds = Rect().also { window.getWindowVisibleDisplayFrame(it) }
        val startTime = SystemClock.elapsedRealtime()

        val root = serializeView(window, 0, startTime, density) ?: emptyMap()

        val result = mutableMapOf<String, Any>(
            "timestamp" to ts,
            "screen" to mapOf(
                "width" to (bounds.width() / density).roundToInt(),
                "height" to (bounds.height() / density).roundToInt(),
                "scale" to displayMetrics.density
            ),
            "root" to root
        )

        ReplayOrchestrator.shared?.currentScreenName?.let {
            result["screenName"] = it
        }

        return result
    }

    private fun serializeView(view: View, depth: Int, startTime: Long, density: Float): Map<String, Any>? {
        if (depth > maxDepth) return null
        if (SystemClock.elapsedRealtime() - startTime > timeBudgetMs) {
            return mapOf("type" to simpleNameOf(view), "bailout" to true)
        }
        if (depth > 0 && (!view.isShown || view.alpha <= 0.01f || view.width <= 0 || view.height <= 0)) {
            return null
        }

        val node = mutableMapOf<String, Any>()
        node["type"] = simpleNameOf(view)

        val location = IntArray(2)
        view.getLocationInWindow(location)
        node["frame"] = mapOf(
            "x" to location[0] / density,
            "y" to location[1] / density,
            "w" to view.width / density,
            "h" to view.height / density
        )

        if (!view.isShown) node["hidden"] = true
        if (view.alpha < 1.0f) node["alpha"] = view.alpha

        // Get accessibility identifier / test ID
        view.contentDescription?.toString()?.takeIf { it.isNotEmpty() }?.let {
            node["testID"] = it
        }

        // Check for React Native nativeID
        try {
            val nativeId = optionalResourceTag(view, "view_tag_native_id", "com.facebook.react") as? String
            if (!nativeId.isNullOrEmpty()) {
                node["testID"] = nativeId
            }
        } catch (_: Exception) { }

        // Resolved once: isSensitive does a resource-tag lookup and a string
        // allocation, and it used to be asked twice for every view in the tree.
        val sensitive = isSensitive(view)
        if (sensitive) node["masked"] = true

        if (includeVisualProperties) {
            view.background?.let { bg ->
                // Try to get background color
                try {
                    val colorDrawable = bg as? android.graphics.drawable.ColorDrawable
                    colorDrawable?.color?.let { color ->
                        node["bg"] = String.format("#%06X", 0xFFFFFF and color)
                    }
                } catch (_: Exception) { }
            }
        }

        if (includeTextContent) {
            when (view) {
                is TextView -> {
                    val text = view.text?.toString() ?: ""
                    node["text"] = if (sensitive) "***" else maskText(text)
                    node["textLength"] = text.length

                    if (view is EditText) {
                        view.hint?.toString()?.let { node["placeholder"] = it }
                    }
                }
            }
        }

        if (isInteractive(view)) {
            node["interactive"] = true

            when (view) {
                is Button -> {
                    node["buttonTitle"] = view.text?.toString() ?: ""
                    node["enabled"] = view.isEnabled
                }
                is CompoundButton -> {
                    node["checked"] = view.isChecked
                    node["enabled"] = view.isEnabled
                }
            }

            if (view.isEnabled) {
                node["enabled"] = true
            } else {
                node["enabled"] = false
            }
        }

        if (view is ScrollView || view is HorizontalScrollView) {
            node["scrollEnabled"] = true
            node["contentOffset"] = mapOf<String, Any>(
                "x" to (((view as? HorizontalScrollView)?.scrollX ?: (view as? ScrollView)?.scrollX ?: 0) / density),
                "y" to (((view as? HorizontalScrollView)?.scrollY ?: (view as? ScrollView)?.scrollY ?: 0) / density)
            )
        }

        if (view is ImageView) {
            node["hasImage"] = view.drawable != null
        }

        // Process children
        if (view is ViewGroup) {
            if (depth >= maxDepth && view.childCount > 0) {
                // Say so rather than silently presenting a leaf. Budget
                // exhaustion already marks itself; depth truncation did not,
                // which is why it went unnoticed.
                node["truncated"] = true
            }
            val children = mutableListOf<Map<String, Any>>()
            for (i in 0 until view.childCount) {
                val child = view.getChildAt(i)
                if (child.isShown && child.alpha > 0.01f) {
                    serializeView(child, depth + 1, startTime, density)?.let {
                        children.add(it)
                    }
                }
            }
            if (children.isNotEmpty()) {
                node["children"] = children
            }
        }

        return node
    }

    private fun isSensitive(view: View): Boolean {
        if (view.contentDescription?.toString() == "rejourney_occlude") return true

        try {
            // Check for React Native accessibility hint tag
            val hint = optionalResourceTag(view, "accessibility_hint", "com.facebook.react") as? String
            if (hint == "rejourney_occlude") return true
        } catch (_: Exception) {
            // Fail closed. This lookup is how a caller marks a view for
            // occlusion through React Native's accessibility hint, so a failure
            // here means we cannot tell whether they asked for it. Swallowing
            // the error and continuing would answer "not sensitive" and put the
            // content in the replay -- the wrong direction for a privacy check.
            return true
        }

        if (view is EditText) {
            val inputType = view.inputType
            // Check for password input types
            if (inputType and android.text.InputType.TYPE_TEXT_VARIATION_PASSWORD != 0 ||
                inputType and android.text.InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD != 0 ||
                inputType and android.text.InputType.TYPE_NUMBER_VARIATION_PASSWORD != 0) {
                return true
            }
            return ReplayOrchestrator.shared?.maskTextInputsByDefault ?: true
        }
        if ((ReplayOrchestrator.shared?.maskTextInputsByDefault ?: true) && isTextInputClass(view)) return true
        return false
    }

    private fun isTextInputClass(view: View): Boolean {
        val className = simpleNameOf(view)
        return className == "ReactEditText" ||
            className == "RCTEditText" ||
            className.contains("TextInput", ignoreCase = true)
    }

    /// Resource ids are fixed for the life of the process, but getIdentifier
    /// resolves them by string every call. This runs per view per scan, and on
    /// an app without React on the classpath every one of those lookups fails,
    /// so the misses are cached as absent too.
    private val resourceIdCache = java.util.concurrent.ConcurrentHashMap<String, Int>()

    /// Class.getSimpleName() derives the name from the binary name and allocates
    /// on each call, and this is asked two or three times per view on every scan
    /// of the whole tree. An app has a small, fixed set of view classes, so the
    /// cache stays tiny and never needs eviction.
    private val simpleNameCache = java.util.concurrent.ConcurrentHashMap<Class<*>, String>()

    private fun simpleNameOf(view: View): String =
        simpleNameCache.getOrPut(view.javaClass) { view.javaClass.simpleName }

    private fun optionalResourceTag(view: View, name: String, packageName: String): Any? {
        val identifier = resourceIdCache.getOrPut("$packageName:$name") {
            view.resources.getIdentifier(name, "id", packageName)
        }
        return if (identifier == 0) null else view.getTag(identifier)
    }

    private fun isInteractive(view: View): Boolean {
        return view is Button ||
               view is EditText ||
               view is CheckBox ||
               view is RadioButton ||
               view is Switch ||
               view is SeekBar ||
               view is Spinner ||
               view.isClickable
    }

    private fun maskText(text: String): String {
        return if (text.length > 100) text.take(100) + "..." else text
    }
}
